### PR TITLE
feat: timm training script Core API example as a blog post.

### PR DESCRIPTION
**Pull request by coreystatendet**
## Description

An example / blog post, converting the `timm` training script from [here](https://github.com/rwightman/pytorch-image-models) to use Core API in steps.

Not intended for merge at this point, only review/comments -- maybe this should be in a separate repo?

Depends on unmerged Core API and Native DTrain features.

## Test Plan

N / A

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: examples/core_api/timm-post/post.md [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+```
+torch.distributed.init_process_group(backend="nccl", init_method="env://")
+```
+[init_process_group](https://pytorch.org/docs/stable/distributed.html#torch.distributed.init_process_group) will then pull information from environment variables about how to connect with other processes and coordinate distributed training.
+
+When we set `slots_per_trial` to something bigger than 1 in our experiment configuration, the Determined master will take care of starting up multiple copies of our entrypoint in containers distributed appropriately across our cluster -- we just need to set those environment variables correctly in each container.  In 0.17.XX, we've provided a [launch script](https://github.com/determined-ai/determined/tree/master/harness/determined/launch) that wraps an entrypoint and handles this environment setup, along with similar scripts for DeepSpeed and Horovod.  Our main step toward distributed training is to modify our `entrypoint` with the appropriate launch script.  While we're at it, we'll bump `slots_per_trial` to 2 to tell the Determined master we want to use 2 GPUs:
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:42]: Gene has requested that we stick to the phase "launcher" instead of "launch layer" or "launch script", which I agree with, so we should probably change that throughout.

File: examples/core_api/timm-post/post.md [hunk 2]
```diff
@@ -0,0 +1,286 @@
+(This post references unreleased features, namely https://github.com/determined-ai/determined/pull/3859 and https://github.com/determined-ai/determined/pull/3807.  To run examples, you'll need to create a local branch with those two PRs merged.)
+
+(TODO: Replace github references with doc references where possible.)
+
+(TODO: Are the initial paragraphs the right story to tell about Core API?  Could completely replace them.)
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:5]: Agreed, we should tell a better story for marketing reasons.  I'll write something up here shortly.

File: examples/core_api/timm-post/post.md [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+The `DistributedContext` we create here provides simple synchronization primitives used to coordinate Core API functionality across workers.  If you're running with one of our supported launchers, you can use the correspondng `from_` helper function to create one from the corresponding environment variables -- otherwise, you can [initialize one manually](https://github.com/determined-ai/determined/blob/master/harness/determined/_core/_distributed.py).
+
+Reporting a dictionary of metrics is now a single API call:
+```
+# The original script already takes care of aggregating metrics across workers.
+train_metrics = train_one_epoch(...)
+
+# Metrics can only be reported on rank 0, to avoid duplicate reports.
+if args.rank == 0:
+    core_context.train.report_training_metrics(latest_batch=latest_batch, metrics=train_metrics)
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:98]: latest_batch is now called steps_completed throughout the api

File: examples/core_api/timm-post/post.md [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    def _save(self, save_path, epoch, metric=None):
+        super()._save(save_path, epoch, metric)
+        checkpoint_metadata = {
+            "latest_batch": self.epoch_length * (epoch + 1),
+            "trial_id": self.trial_id,
+        }
+        with core_context.checkpoint.store_path(checkpoint_metadata) as (path, _):
+            shutil.copy2(save_path, path.joinpath("data"))
+```
+The essential Core API call here is `core_context.checkpoint.store_path`, which returns a context manager that:
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:149]: have you considered using core_context.checkpoint.upload() instead of store_path?  It's quite a bit more intuitive, and for medium-to-small checkpoints there's no real benefit to store_path.

File: examples/core_api/timm-post/post.md [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+- On entry, finds or creates a directory for us to write checkpoint files to, returning the path and a unique checkpoint ID.
+- On exit, uploads the files to checkpoint storage (if necessary) and notifies the master.
+After running the parent logic to save a checkpoint to `save_path` , we just `shutil.copy2` the full contents of `save_path` to the Determined checkpoint directory without needing to know anything specific about the `timm` checkpoint format.
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:152]: if you can't direct where the checkpoint is saved to, then upload is strictly better than `store_path` anyway.

File: examples/core_api/timm-post/post.md [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+```
+saver = DeterminedCheckpointSaver(...)
+```
+And we can now see that checkpoints are being saved to storage:
+< TODO: Checkpoint image >
+
+Restoring from a checkpoint is only slightly more complicated:
+```
+if latest_checkpoint is not None:
+    restore_path_context = core_context.checkpoint.restore_path(latest_checkpoint)
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:164]: same thing, consider using .download instead of .restore_path

File: examples/core_api/timm-post/post.md [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+Over time, we've realized that ML teams often have extensive script libraries and custom execution harnesses of their own that make this porting process non-trivial.  To address this, we designed *Core API*: an API with fewer built-in assumptions that can be plugged into existing code with minimal fuss.  We're releasing it as part of Determined 0.17.XX.
+
+In this post, we'll walk through converting an existing training script to use the full set of Core API features.  We'll proceed in steps, to illustrate how Core API enables incremental adoption of features.  And instead of working with a toy example, we'll instead use the training script from Ross Wightman's popular `timm`  library for image models.  TODO: More background on `timm`?
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:11]: For the toy example, you might as well link to the new official core api tutorial, which just landed this week.  See docs/training-apis/core-api/*

File: examples/core_api/timm-post/post.md [hunk 8]
```diff
@@ -0,0 +1,286 @@
...
+
+(TODO: Replace github references with doc references where possible.)
+
+(TODO: Are the initial paragraphs the right story to tell about Core API?  Could completely replace them.)
+
+Users of Determined have always faced a small but annoying barrier to entry - porting their model training scripts to use one of Determined's Trial APIs.  Similar to popular libraries like PyTorch Lightning, our Trial APIs provide user-definable hooks for each step in a standard model training workflow.  Our behind-the-scenes execution harness then handles the messy details of providing scalable, distributed, preemptible training with full access to features like hyperparameter search and checkpoint management.
+
+Over time, we've realized that ML teams often have extensive script libraries and custom execution harnesses of their own that make this porting process non-trivial.  To address this, we designed *Core API*: an API with fewer built-in assumptions that can be plugged into existing code with minimal fuss.  We're releasing it as part of Determined 0.17.XX.
```
Comment (rb-determined-ai) [inline examples/core_api/timm-post/post.md:9]: Let's couch this as a "huge improvement" rather than as a "big bugfix".  And that's not just marketing speak either... I don't know of any of our competitors who offer the level of platform integration that core api offers.  Things like out-of-the-box distributed communication primitives and distributed preemption via a click in the webui are, afaik, unique to us.

------

Until now, Determined has offered scheduling, distributed training, metrics reporting, checkpoint tracking, and hyperparameter search to PyTorch and Keras models.  Users only had to implement the `PyTorchTrial` or `TFKerasTrial` interface to integrate with the platform.

Still, we know many ML teams have extensive script libraries and custom training loops of their own.  Ideally, they would like to access the benefits of the Determined platform without refactoring their code as a Trial class.

So we developed the Determined Core API.  Released in Determined version 0.18.0, the Core API allows users to fully integrate arbitrary models and training loops into the Determined platform.  With minimal fuss, you can now imbue any training loop with all of the features offered by the Trial APIs.
